### ISSUE TITLE
kubelet: panic runtime error: invalid memory address or nil pointer dereference in net.IsIPv4String

### ISSUE THREAD

**Issue by mborsz** [kubernetes/kubernetes#102806]
#### What happened:
In https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/1401222770964041728 one of kubelets panicked with error like:
```
I0605 19:32:06.921560    2102 kubelet.go:1976] "SyncLoop (PLEG): event for pod" pod="test-b7x1na-8/small-deployment-245-84cdf6f8f5-znc2g" event=&{ID:3e8a37c1-dbb0-434f-b51a-4af2d5f7b25e Type:ContainerDied Data:4085c130ddda52ae1d0655ecb410ff32f452f7f53cceb368e73d42dbef6fbbb5}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x55db470207b9]
goroutine 812 [running]:
net.ParseIP(0x0, 0xa, 0x55db4b52cca0, 0xc0011e0d01, 0xc003e658b0)
	/usr/local/go/src/net/ip.go:680 +0x39
k8s.io/kubernetes/vendor/k8s.io/utils/net.IsIPv4String(0x0, 0xa, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/utils/net/net.go:147 +0x3b
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).convertStatusToAPIStatus(0xc000691500, 0xc001765c00, 0xc001059a80, 0xc00027eba0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_pods.go:1591 +0x187
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).generateAPIPodStatus(0xc000691500, 0xc001765c00, 0xc001059a80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_pods.go:1516 +0x1d1
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).cleanUpContainersInPod(0xc000691500, 0xc0027e9650, 0x24, 0xc001260f00, 0x40)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2289 +0x150
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoopIteration(0xc000691500, 0xc000bf3f20, 0x55db4bdbc770, 0xc000691500, 0xc0018476e0, 0xc001847740, 0xc000f485a0, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1986 +0x6ae
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoop(0xc000691500, 0xc000bf3f20, 0x55db4bdbc770, 0xc000691500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1886 +0x35b
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run(0xc000691500, 0xc000bf3f20)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1466 +0x28c
created by k8s.io/kubernetes/cmd/kubelet/app.startKubelet
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:1200 +0x57
kubelet.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
kubelet.service: Failed with result 'exit-code'.
kubelet.service: Consumed 11min 36.305s CPU time
kubelet.service: Service RestartSec=10s expired, scheduling restart.
kubelet.service: Scheduled restart job, restart counter is at 1.


```

https://storage.googleapis.com/sig-scalability-logs/ci-kubernetes-e2e-gce-scale-performance/1401222770964041728/gce-scale-cluster-minion-group-1-v2lp/kubelet.log
#### What you expected to happen:

kubelet doesn't panic

#### How to reproduce it (as minimally and precisely as possible):
I don't know unfortunately, but it's a second time I'm seeing this kind of issue in our tests. I hope stacktrace will be helpful to debug this.

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by ehashman** [comment]
Looks like this affects master (1.22).

/triage accepted
/priority critical-urgent

**Comment by ehashman** [comment]
https://github.com/kubernetes/utils/blob/2afb4311ab10fb869c5d8c260b5e92b4bbde7f80/net/net.go#L145-L149

I'm not sure how that would be throwing an NPE.

https://golang.org/pkg/net/#ParseIP

**Comment by cndoit18** [comment]
> https://github.com/kubernetes/utils/blob/2afb4311ab10fb869c5d8c260b5e92b4bbde7f80/net/net.go#L145-L149
> 
> I'm not sure how that would be throwing an NPE.
> 
> https://golang.org/pkg/net/#ParseIP

I can reproduce it, but I can't find the source

**Comment by ehashman** [comment]
https://github.com/kubernetes/kubernetes/blame/3dd561ade4cfde179af0e17b029bff27c537a534/pkg/kubelet/kubelet_pods.go#L1581-L1589

/cc @danwinship 
/sig network

**Comment by endocrimes** [comment]
@cndoit18 How'd you manage to repro it? - I can't figure out how to get a string into such a broken state off the top of my head

**Comment by ehashman** [comment]
https://golang.org/src/net/ip.go?s=15720:15745#L668 for the upstream code.

**Comment by cndoit18** [comment]
```
        s := "1.1.1.1"
	point := &s
	point = nil
	net.ParseIP(*point)
net.
```

**Comment by ehashman** [comment]
This is the line throwing the NPE: https://github.com/kubernetes/kubernetes/blob/3dd561ade4cfde179af0e17b029bff27c537a534/pkg/kubelet/kubelet_pods.go#L1591

I don't know how we'd call that with a null pointer, happy for others to dig :)

**Comment by endocrimes** [comment]
@cndoit18 hmm - that's not quite the same crash  - That fails when dereferencing the pointer, not when trying to access the string in the loop

**Comment by cndoit18** [comment]
> @cndoit18 hmm - that's not quite the same crash - That fails when dereferencing the pointer, not when trying to access the string in the loop

```
	s := "1.1.1.1"
	vs := reflect.ValueOf(s)

	tmp := ""
	point := &tmp
	point = nil
	vs.SetString(*point)

	net.ParseIP(s)
```

**Comment by endocrimes** [comment]
@ehashman Yeah - It _seems_ like this should only occur if somehow the string had length but a borked data pointer, and I'm not quite sure how that can happen naturally in the code that I'm reading in the pod deregistration flow

**Comment by endocrimes** [comment]
@cndoit18 That also crashes when deferencing *point rather than ever making it to net.ParseIP. I don't think there's a way you can break the string len like this easily because SetString under the hood derefs the value pointer and assigns the new value.

**Comment by aojea** [comment]
https://github.com/kubernetes/kubernetes/pull/102817

**Comment by aojea** [comment]
/close
This should be fixed by #103307
